### ISSUE TITLE
Metrics in different gauges and labels such as keyspace and table names

### ISSUE THREAD

**Issue by ffarozan** [criteo/cassandra_exporter#3]
Current implementation is such that all metrics are labels in a gauge named as "cassandra_stats".

I think, a better design would be similar to what JMX prometheus exporter does. Gauges are separated out. And labels can be configurable such as keyspace name or table name etc.

Any thoughts?

**Comment by erebe** [comment]
Hello @ffarozan,
Can you give me an example of what a cassandra_stats{...} would look like to you ?

**Comment by ffarozan** [comment]
Hi @erebe

For example let us take readlatency at column family or table level.
org:apache:cassandra:metrics:table:readlatency:count{keyspace="system", table="hints"}
org:apache:cassandra:metrics:table:readlatency:count{keyspace="kairosdb", table="data_points}

There is no advantage of making different cassandra metrics as labels under a global "cassandra_stats" metric. IMO the cassandra metrics can be grouped under its own metrics, and use labels for keyspace, table, etc.

**Comment by ffarozan** [comment]
@erebe The current implementation has the bug of not sending the individual keyspace or table metrics out. It may be overwriting the same metric and sending just the last one it is reading.

**Comment by erebe** [comment]
Regarding the proposition for
`org:apache:cassandra:metrics:table:readlatency:count{keyspace="system", table="hints"}`

The choice of the global namespace has been made in order to **help discoverability of metrics**. 
We were tired of trying to remember which from {scope,type,name,keyspace, ...} of the jmx object name was the actual order/meaning of the jmx path. With this design choice when you want to look at a specific metric that you are not usually displaying (i.e: when you want to debug) you don't have any hassle to find it.

Regarding introducing specific labels `{keyspace=, table=}` instead a single global one, we wanted to do that at first but this introduce semantics regarding the metrics, and this semantics is specific to each metric. For example `org:apache:cassandra:metrics:connection:xxx` contains subpath for each ip address, so you cannot use the label keyspace for every same path depth, you have to encode in the code that if the metrics name is `org:apache:cassandra:metrics:connection`you need to use the label ip instead of keyspace. Those specifics case for each metrics make the code less generic and so will require more works if evolutions happens.
We thought also of doing specific rules like that for commons metrics like tables or keyspaces and fallback to the global namespace, but making a distinction seems odd to us.

And anyway we are reintroducing semantics in the metrics with our prometheus rules in order to aggregate metrics for a whole cluster. 
```
instance:cassandra_connections_gossipmessagedropped:value = sum(
  max(
     cassandra_stats{name=~"org:apache:cassandra:metrics:connection:.*:gossipmessagedroppedtasks:value"}
  ) by (cluster, dc, exported_instance, name)
) by (cluster, dc, exported_instance)
```


Regarding 
> @erebe The current implementation has the bug of not sending the individual keyspace or table metrics out. It may be overwriting the same metric and sending just the last one it is reading.

Can you provide an example, I am not sure to understand

**Comment by ffarozan** [comment]
Discoverability of metrics should not be of concern. If someone is interested in making sense of the metrics, they should understand the details.

>Regarding introducing specific labels {keyspace=, table=} instead a single global one, we wanted to do that at first but this introduce semantics regarding the metrics, and this semantics is specific to each metric. For example org:apache:cassandra:metrics:connection:xxx contains subpath for each ip address, so you cannot use the label keyspace for every same path depth, you have to encode in the code that if the metrics name is org:apache:cassandra:metrics:connectionyou need to use the label ip instead of keyspace. Those specifics case for each metrics make the code less generic and so will require more works if evolutions happens.

The labels will, of course, need to be different for different metrics. As you said, for connection metrics, it would be ip address. And for keyspace metrics, it would be keyspace. For table metrics, it would be table and keyspace. If you agree this is useful and needed, I can take a stab at it.
I liked the design of JMX exporter from that aspect. It lets users choose how to convert each metric, instead of predefining the structure. For ex: my prometheus config yaml looks like:

```
rules:
- pattern: 'org.apache.cassandra.metrics<type=(ClientRequest), scope=(Read|Write|RangeSlice), name=(Latency|TotalLatency)><>(Count|OneMinuteRate|FiveMinuteRate)'
  name: cassandra_$1_$2_$3_$4
- pattern: 'org.apache.cassandra.metrics<type=(Storage), name=(Load)><>(Count)'
   name: cassandra_$1_$2_$3
```

The main reason I am looking at criteo is it being standalone and the scraping interval customization per metric. If you agree we should enhance criteo to handle scenarios like above, I can take a stab at it.

> Can you provide an example, I am not sure to understand

I just now figured out the individual keyspace or table level metrics are not scraped by criteo. I am referring to the metrics like:
```
org.apache.cassandra.metrics:type=ColumnFamily,keyspace=system,scope=hints,name=LiveSSTableCount
```
In that case, my statement about the bug is wrong. It is not a bug in the current implementation. But my requirement is to scrape these individual metrics as well.

**Comment by erebe** [comment]
If you are willing to submit a PR to add semantic's labels to all metrics, feel free to do it. But be aware that this is not a commitment of myself to merge the future PR, we will discuss based upon it.
In any case, if you do it you will be able to use it :) and you can feel free to ping me if you have some trouble with the code.

Regarding 

```
org.apache.cassandra.metrics:type=ColumnFamily,keyspace=system,scope=hints,name=LiveSSTableCount
```
If you are using the config file provided in the repository this is expected that those metrics to not be parsed, as there is a blacklist for the `columnfamily` and `system` path
```yaml
blacklist: 
   # columnfamily is an alias for Table metrics
   # https://github.com/apache/cassandra/blob/8b3a60b9a7dbefeecc06bace617279612ec7092d/src/java/org/apache/cassandra/metrics/TableMetrics.java#L162
   - org:apache:cassandra:metrics:columnfamily:.*

   # Should we export metrics for system keyspaces/tables ?
   - org:apache:cassandra:metrics:[^:]+:system[^:]*:.*
```
If you want to scrap those metrics, just remove those 2 blacklist in your config file.

P.S: Be aware that the columnfamily and the table metrics are the same, this is due to historic reason where columnfamily was used to refer to the same concept to what is now called table in Cassandra 3.x. I don't know in which version the rename happened so maybe you don't have the table path in Cassandra 2.x.
Anyway, if you are withelisting the columnfamilly path, put a blacklist for the table path in order to not get duplicate metrics when you switch to cassandra 3.x
```
org.apache.cassandra.metrics:type=ColumnFamily,name=LiveSSTableCount
is equal to in cassandra 3.x
org.apache.cassandra.metrics:type=Table,name=LiveSSTableCount
```

**Comment by ffarozan** [comment]
Yes. I know about the current blacklisting of the columnfamily metrics in the config.yml. 

And I am using 2.x. So there is no *table* metrics in there. Probably those metrics were added in 3.x.
